src/internal/collection/collection_test.go

```diff
@@ -1,699 +1,1352 @@
+
+					createKeys := []string{"0", "6", "7", "9", "3", "8", "4", "1", "2", "5"}
+					for _, k := range createKeys {
+						err := writer(context.Background(), func(rw col.ReadWriteCollection) error {
+							testProto := &col.TestItem{ID: k, Value: originalValue}
+							return rw.Create(k, testProto)
+						})
+						require.NoError(t, err)
+					}
+
+					checkDefaultCollection(t, createCol, RowDiff{})
+					testSort(t, createCol, col.SortByCreateRevision, createKeys)
+				})
+
+				subsuite.Run("UpdatedAt", func(t *testing.T) {
+					// Create a new collection that we can modify to get a different ordering here
+					reader, writer := newCollection(context.Background(), t)
+					modRead := reader(context.Background())
+					require.NoError(suite, writer(context.Background(), populateCollection))
+
+					modKeys := []string{"5", "7", "2", "9", "1", "0", "8", "4", "3", "6"}
+					for _, k := range modKeys {
+						err := writer(context.Background(), func(rw col.ReadWriteCollection) error {
+							testProto := &col.TestItem{}
+							if err := rw.Update(k, testProto, func() error {
+								testProto.Value = changedValue
+								return nil
+							}); err != nil {
+								return err
+							}
+							return nil
+						})
+						require.NoError(t, err)
+					}
+					testSort(t, modRead, col.SortByModRevision, modKeys)
+				})
+
+				subsuite.Run("Key", func(t *testing.T) {
+					testSort(t, defaultRead, col.SortByKey, idRange(0, 10))
+				})
+			})
+
+			subsuite.Run("ErrorInCallback", func(t *testing.T) {
+				t.Parallel()
+				count := 0
+				testProto := &col.TestItem{}
+				err := defaultRead.List(testProto, col.DefaultOptions(), func() error {
+					count++
+					return &TestError{}
+				})
+				require.YesError(t, err)
+				require.True(t, errors.Is(err, TestError{}), "Incorrect error: %v", err)
+				require.Equal(t, 1, count, "List callback was called multiple times despite erroring")
+			})
+
+			subsuite.Run("ErrBreak", func(t *testing.T) {
+				t.Parallel()
+				count := 0
+				testProto := &col.TestItem{}
+				err := defaultRead.List(testProto, col.DefaultOptions(), func() error {
+					count++
+					return errutil.ErrBreak
+				})
+				require.NoError(t, err)
+				require.Equal(t, 1, count, "List callback was called multiple times despite aborting")
+			})
+		})
+
+		suite.Run("Count", func(subsuite *testing.T) {
+			subsuite.Parallel()
+			subsuite.Run("Success", func(t *testing.T) {
+				t.Parallel()
+				count, err := defaultRead.Count()
+				require.NoError(t, err)
+				require.Equal(t, int64(10), count)
+
+				count, err = emptyRead.Count()
+				require.NoError(t, err)
+				require.Equal(t, int64(0), count)
+			})
+		})
+
+		// 'Watch' functions with callbacks don't have a good hook to trigger
+		// events. This is a bit racy, but we can sleep a bit to let the listen get
+		// started, then trigger some writes.
+		asyncWrite := func(t *testing.T, writer WriteCallback, cbs ...func(rw col.ReadWriteCollection) error) {
```

Comment:
This pattern seems like it will become a source of flaky tests.  I think the way to do this is set up a listener that copies to a channel.  Then write in a goroutine, and receive in the main.  You can also select from a context so that it times out in the failing case.

Reply:
I think I have an alternative here that will be a bit more readable and concise, will upload it tonight.  Going to merge back all the changes to `src/internal/collection` that came up in the integration branch.